polaris-react/src/components/IndexTable/IndexTable.tsx

```diff
@@ -863,6 +877,8 @@ function IndexTableBase({
     index: number,
     direction: IndexTableSortDirection,
   ) {
+    renderAfterSelectEvent.current = false;
+    hasSelected.current = false;
     lastSortedColumnIndex.current = sortColumnIndex;
```

Comment:
Without the `hasSelected` logic, the flicker issue is mostly resolved, except for the first time a row or "select all" is selected.